Feat(Problem): 북마크한 문제 목록 조회 API 구현

morenow98 opened the pull request:
## PR 변경된 내용
### 북마크한 문제 목록 조회 API 구현
1. 일부 검증의 책임을 각 도메인으로 이전, 이후 검증만을 책임지는 Validator 혹은 DTO의 Valid 어노테이션 활용으로 리팩토링 가능성 있음.
2. examId의 존재 유무에 따라 문제를 찾는 로직이 복잡함에 따라 관심사의 분리가 필요. 따라서 domain 모듈에 해당 로직을 담당하는 Service 클래스 구현

### 테스트에서 있었던 이슈
1. TestFixtures 플러그인 적용
- api 모듈에 있던 fixture 클래스들을 domain으로 옮기고, 모든 모듈에서 활용
- 테스트 파일들은 jar 파일에 포함되지 않으므로 모듈 간 공유가 불가능
- gradle에서 제공하는 java-test-fixtures 플러그인을 활용해 모듈간 fixture 클래스 공유
- 참조 : https://toss.tech/article/how-to-manage-test-dependency-in-gradle
2. Repository 테스트 구현
- 북마크 목록 조회는 복잡한 쿼리가 나가므로 repository 테스트를 구현
- test를 위한 설정 파일 추가, H2 database 활용
- domain 모듈에는 ComponentScan이 없어 이를 위해 같은 경로에 SpringBootApplication을 위한 클래스 생성
  - 참조 : https://velog.io/@jonghyun3668/SpringBoot-%EB%A9%80%ED%8B%B0-%EB%AA%A8%EB%93%88-%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B8%EC%97%90%EC%84%9C-%ED%85%8C%EC%8A%A4%ED%8A%B8-%ED%99%98%EA%B2%BD-%EC%84%A4%EC%A0%95%ED%95%98%EA%B8%B0
- 북마크 목록 조회 repository에 EntityManager 사용
- EH2 database의 예약어 문제로 xam의 year 컬럼명 임시 변경
- DataJpaTest 에 대한 참고 글 : https://velog.io/@xogml951/JUnit5%EA%B3%BC-Spring-boot-%ED%85%8C%EC%8A%A4%ED%8A%B8-%EC%BD%94%EB%93%9C-%EC%9E%91%EC%84%B1%EB%B2%95-Repository-Layer-%EB%8B%A8%EC%9C%84%ED%85%8C%EC%8A%A4%ED%8A%B8-%EB%B0%A9%EB%B2%952

## 추가 내용
- Problem에서 Bookmark를 참조하는 경우가 없어 양방향 매핑을 하지 않도록 변경
- Problem에서 생성 시 createdAt 값이 생성되지 않는 현상 수정
- 문제 풀이에서 과목당 문제 수가 최소한 1개는 되어야 하는 것으로 수정
- Jira에서 Github로 이전하면서 pull request template 변경
- 테스트 설명을 구체적으로 변경

## 참조
Closes #19

morenow98 committed:
Feat(Problem): Problem에서 Bookmark를 참조하는 경우가 없어 양방향 매핑을 하지 않도록 변경

morenow98 committed:
Fix(Problem): Problem에서 생성 시 createdAt 값이 생성되지 않는 현상 수정

- @EntityListeners(AuditingEntityListener.class) 추가

morenow98 committed:
Feat(Problem): 북마크 전체 목록 조회 API 구현 (미완성)

morenow98 committed:
Test(Problem): 북마크 전체 목록 조회 API 테스트 구현 (미완성)

morenow98 committed:
Feat(Problem): 과목당 문제 최소 개수 1개로 수정

morenow98 committed:
Refactor(Problem): 검증의 책임을 각 도메인으로 이전

morenow98 committed:
Refactor(Certificate): 자격증 현재 상태를 찾지 못했을 때, 에러를 세분화

자격증을 찾지 못했습니다. -> 자격증 상태가 존재하지 않습니다.

morenow98 committed:
Feat(Problem): 북마크한 문제 목록 조회 API 세부 구현

- examId 존재 유무에 따라 문제를 찾아오는 로직을 domain 모듈의 problem 패키지에서 담당하도록 함
- 리팩토링해야 할 부분에 많은 TODO 표시 처리함

morenow98 committed:
Test(Problem): 북마크한 문제 목록 조회 테스트 구현

- api 모듈의 usecase, domain 모듈의 service 모두 테스트 구현

morenow98 committed:
Feat(*): Jira를 사용하지 않음에 따라 참조에 예시로 있던 Jira issue 문구 삭제

morenow98 committed:
Fix(Problem): 수정하지 않았던 테스트 코드 수정

InHyeok-J commented on jabiseo-api/src/main/java/com/jabiseo/problem/usecase/FindBookmarkedProblemsUseCase.java:
```diff
@@ -1,33 +1,54 @@
+
+        Member member = memberRepository.getReferenceById(memberId);
+
+        if (!member.containsCertificate()) {
+            throw new MemberBusinessException(MemberErrorCode.CURRENT_CERTIFICATE_NOT_EXIST);
+        }
```
contains라는 네이밍이 메소드 행위랑 안맞는거 같아서 현재 자격증 상태가 등록돼있는지를 판단하는 네이밍이 적용되면 좋을거같음(ListX)

morenow98 replied:
has 메소드 명으로 바꿨습니다

morenow98 committed:
Refactor(Problem): 헷갈릴 수 있는 메소드명 변경

- CertificateStatus는 List가 아니기 때문에 contains 에서 has 로 네이밍 변경

InHyeok-J commented on jabiseo-api/src/main/java/com/jabiseo/problem/usecase/FindBookmarkedProblemsUseCase.java:
```diff
@@ -1,33 +1,54 @@
+                .map(Problem::getId)
+                .toList();
+
+        List<Problem> problems = bookmarkedProblemService.findBookmarkedProblems(problemIds, examId, subjectIds, pageable);
```
이거 수정 필요

morenow98 replied:
수정해서 올렸습니다

InHyeok-J commented on jabiseo-api/src/main/java/com/jabiseo/problem/usecase/FindBookmarkedProblemsUseCase.java:
```diff
@@ -1,33 +1,54 @@
+    private final MemberRepository memberRepository;
+    private final BookmarkedProblemService bookmarkedProblemService;
+
+
+    public List<FindBookmarkedProblemsResponse> execute(String memberId, Optional<String> examId, List<String> subjectIds, Pageable pageable) {
+
```
해당 메소드의 책임을 정리해봤는데
1. 검증
    - 현재 자격증을 갖고 있는지 확인하고
    - 갖고 있으면 그 자격증 기반으로 입력받은 subjects가 정확한지 검증하고
    - exam까지 정확한지 검증한다
2. 쿼리
    - 이후 examid 여부에 따라
    - 다른 2가지 쿼리를 날린다.
    - if queryDSL이라면 where null인거 자동 체크해줄텐데…
    
      코드에 고민이 있는거같은데 크게 이렇게 2개 뽑아서 한번 분리해보는건 어떨까?

morenow98 replied:
수정해서 올렸습니다.

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/problem/service/BookmarkedProblemService.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final ProblemRepository problemRepository;
+
+    public List<Problem> findBookmarkedProblems(List<String> problemIds, Optional<String> examId, List<String> subjectIds, Pageable pageable) {
+        // examId의 존재 유무에 따라 다른 쿼리를 날려 문제를 가져옴
+        return examId.map(rawExamId ->
+                        problemRepository.findByIdInAndExamIdAndSubjectIdIn(problemIds, rawExamId, subjectIds, pageable))
+                .orElseGet(() -> problemRepository.findByIdInAndSubjectIdIn(problemIds, subjectIds, pageable))
+                .toList();
+    }
```
메소드 examId에 따라서 분기 처리해서 다른 쿼리를 날리는게 핵심 책임인것 같은데

이러면 if가 있는게 직관적으로 좋은 거 같음

morenow98 replied:
수정해서 올렸습니다.

morenow98 replied:
UseCase 에 보면 if(examId.isPresent()) 로 올려놨는데 두 코드 비교해서 어떤 게 나은지 봐봐

```java
if (examId.isPresent()) {
    return problemRepository.findBookmarkedByExamIdAndSubjectIdIn(memberId, examId.get(), subjectIds, pageable)
            .stream()
            .map(FindBookmarkedProblemsResponse::from)
            .toList();
} else {
    return problemRepository.findBookmarkedBySubjectIdIn(memberId, subjectIds, pageable)
            .stream()
            .map(FindBookmarkedProblemsResponse::from)
            .toList();
}
```

```java
List<Problem> problems = examId.map(id ->
                problemRepository.findBookmarkedByExamIdAndSubjectIdIn(memberId, id, subjectIds, pageable))
        .orElseGet(() -> problemRepository.findBookmarkedBySubjectIdIn(memberId, subjectIds, pageable));

return problems.stream()
        .map(FindBookmarkedProblemsResponse::from)
        .toList();
```

InHyeok-J replied:
지금보니 둘다 좋은데? 너가 편한거 선택하면 될듯

morenow98 replied:
아래 걸로 올렸음

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/certificate/domain/Certificate.java:
```diff
@@ -57,4 +60,20 @@ public boolean containsExam(String examId) {
                 .map(Exam::getId)
                 .anyMatch(id -> id.equals(examId));
     }
+
+    public void validateSubjectIds(List<String> subjectIds) {
+        // 자격증에 해당하는 과목들이 모두 있는지 검사
+        subjectIds.forEach(subjectId -> {
+            if (!this.containsSubject(subjectId)) {
+                throw new CertificateBusinessException(CertificateErrorCode.SUBJECT_NOT_FOUND_IN_CERTIFICATE);
+            }
+        });
+    }
+
+    public void validateExamId(Optional<String> examId) {
+        // 자격증에 해당하는 시험이 있는지 검사
+        if (examId.isPresent() && !this.containsExam(examId.get())) {
+            throw new CertificateBusinessException(CertificateErrorCode.EXAM_NOT_FOUND_IN_CERTIFICATE);
+        }
+    }
```
validate가 아니라 contains으로 처리하고 
validate 메소드 안에 2개를 호출하는게 어떄?

```java
public void validateContainsSubjectAndExam(List subjects, Optional exam){
    this.containsSubjects(subjects);
    this.containsExam(exam);
}

private void containsSubjects(){}
private void containsExam(){}
```
2개를 호출하는게 같이 가니깐 하나로 묶어도 좋을거같음

morenow98 replied:
수정해서 올렸습니다.

InHyeok-J replied:
@morenow98 public하게 노출이 안되는건 일단 private 으로 두는건 어떄

morenow98 replied:
수정 완료

InHyeok-J reviewed (changes requested):
다시 리뷰주세요

morenow98 committed:
Refactor(Problem): 멤버가 현재 자격증을 가지고 있지 않는 경우 예외처리를 get 메서드에서 하도록 변경

morenow98 committed:
Feat(Problem): 페이지 크기를 20에서 10으로 변경

- 기본 Pageable을 생성하는 데서 PageRequest를 직접 생성하는 것으로 변경
- Pageable을 컨트롤러가 아닌 UseCase에서 생성

morenow98 committed:
Refactor(*): TestFixtures 플러그인 적용

- api 모듈에 있던 fixture 클래스들을 domain으로 옮기고, 모든 모듈에서 활용
- 테스트 파일들은 jar 파일에 포함되지 않으므로 모듈 간 공유가 불가능
- gradle에서 제공하는 java-test-fixtures 플러그인을 활용해 모듈간 fixture 클래스 공유

morenow98 committed:
Refactor(*): 테스트 설명을 구체적으로 변경

morenow98 committed:
Feat(Problem): 북마크 목록 조회 구현 변경

- 기존에는 멤버의 북마크를 모두 조회해서 북마크를 많이 헀을 경우 문제가 발생
- 데이터베이스에서 직접 join해서 한 번에 가져오는 것으로 변경

morenow98 committed:
Test(Problem): 북마크 목록 조회 repository 테스트 구현

- test를 위한 설정 파일 추가, H2 database 활용
- domain 모듈에는 ComponentScan이 없어 이를 위해 같은 경로에 SpringBootApplication을 위한 클래스 생성
- 북마크 목록 조회 repository에 EntityManager 사용
- EH2 database의 예약어 문제로 xam의 year 컬럼명 임시 변경

morenow98 committed:
Feat(Problem): 과목 당 문제수 검증을 controller에 추가

- API 명세에 따른 단순 입력값 등의 검증은 Controller 와 DTO 에서 하기로 팀원들과 결정
- @Valid 어노테이션으로 입력값 검증을 함에 따라 ExceptionHandler 구현 변경
- COM_001을 invalid request body에서 invalid request로 확장

morenow98 committed:
Chore(*): 클래스 설명을 위한 단순 주석 추가

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/member/domain/Member.java:
```diff
@@ -80,4 +84,11 @@ public Member updateCertificateState(Certificate certificate) {
         bookmarks.add(bookmark);
     }
+
+    public Certificate getCertificateState() {
+        if (this.certificateState == null) {
+            throw new MemberBusinessException(MemberErrorCode.CURRENT_CERTIFICATE_NOT_EXIST);
+        }
+        return certificateState;
+    }
```
Member에 현재 진행중인 자격증이 없으면 예외를 반환시켜야 하나?
이거에 관한 API 문서에 정의는 없는데 FindMyCertificateStateResponse 클래스 보면 of 메소드 안에 null 인경우 null로 리턴하는 코드로 구현중인데 해당 부분 동기화 작업이 필요할듯
그리고 이 코드가 테스트가 통과된건 FindCertificateUseCase에 null 인경우 빈 값을 넣는거를 테스트 안하는데 테스트가 필요해보임.

morenow98 replied:
validate 하는 메소드를 따로 분리해서 구현해서 올렸음

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/member/exception/MemberErrorCode.java:
```diff
@@ -7,6 +7,7 @@
 public enum MemberErrorCode implements ErrorCode {
 
     MEMBER_NOT_FOUND("회원을 찾을 수 없습니다.", "MEM_001", ErrorCode.NOT_FOUND),
+    CURRENT_CERTIFICATE_NOT_EXIST("자격증 상태가 존재하지 않습니다.", "MEM_002", ErrorCode.BAD_REQUEST),
```
해당 코드가 400예외를 갖는 이유가 뭐지? client의 예외는 아닌거같은데

morenow98 replied:
회의 결과 보내면 안되는 상황에서만 나는 오류이므로 400이 맞는 것으로~

InHyeok-J commented on jabiseo-domain/src/test/java/com/jabiseo/problem/domain/ProblemRepositoryTest.java:
```diff
@@ -0,0 +1,86 @@
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Test
+    @DisplayName("북마크 목록 조회 JPA 테스트")
+    void givenProblemConditions_whenFindingBookmarkedProblems_thenFindBookmarkedProblems() {
```
해당 JPA 쿼리를 직접 jpql로 짜서 테스트 하는건 좋은듯
근데  Displayname으로는 무엇을 테스트하려고 하는지 알수가 없는거같아 이거 구체화 가능?

morenow98 replied:
굳굳 수정했어

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/certificate/domain/Certificate.java:
```diff
@@ -57,4 +60,25 @@ public boolean containsExam(String examId) {
+        }
+    }
+
+    public void validateAndSubjectIds(Optional<String> examId, List<String> subjectIds) {
+        validateExamId(examId);
+        validateSubjectIds(subjectIds);
+    }
```
validateAndSubjectIds메소드에서 exam 이 파라미터가 있는데 빠진거같음 오타인거같은데?

morenow98 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 정신이 없었나보다...

InHyeok-J commented on jabiseo-api/src/test/java/com/jabiseo/problem/usecase/FindBookmarkedProblemsUseCaseTest.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Mock
+    ProblemRepository problemRepository;
+
+    @Test
+    @DisplayName("북마크 목록 조회를 성공한다.")
+    void givenProblemConditions_whenFindingBookmarkedProblems_thenFindBookmarkedProblems() {
+        //given
+        String memberId = "1";
```
음 해당 테스트코드를 열심히 짰는데 실패 케이스에 대한건 Usecase에서 테스트를 안하나? 
안하면 Certificate 의 도메인 Pojo 테스트라도 추가하면 좋을듯
그리고 성공 케이스에 대해 많이 짰는데 실패 케이스를 먼저 테스트해서 짜는건 어때?

morenow98 replied:
이 유스케이스에서 경우의 수는 다음 경우가 있는데,
1. examId 포함한 성공
2. examId 미포함 성공
3. 멤버가 현재 자격증을 가지고 있지 않을 경우
4. examId와 subjectId가 올바르지 않을 경우
5. repository에서 쿼리가 제대로 동작하지 않을 경우

맨 처음에는 1,3,4 번에 대한 테스트를 usecase에서 짰어. 근데 구현이 변경되면서 1번만 남았고, 5번은 추가로 @DataJpaTest 를 통해서 구현했어.

2,3,4번이 구현되어 있지 않아서 추가로 구현해서 지금 올렸어.
2,3번은 그대로 UseCase 에 추가로 구현했고, 4번의 검증 책임을 Certificate 클래스로 옮긴 상태라서 테스트코드도 그 메소드를 테스트했어

morenow98 replied:
완전 깜빡한 부분인데 떙큐

InHyeok-J reviewed (changes requested):
TestFixtures는 처음보는데 좋네요 GOOD
댓글 단거 몇개 있는데 이거 내일 얘기해보자

InHyeok-J commented:
@morenow98 usecase패키지로 남아있는데 application으로 패키지 네이밍 통일해서 올려줘

morenow98 committed:
Merge branch 'refs/heads/develop' into feature/19-find-bookmark-list

morenow98 committed:
Chore(Problem): 테스트 설명을 구체적으로 변경

morenow98 committed:
Refactor(Problem): 외부에 노출되지 않는 메소드를 private으로 변경

- 관련된 메소드에 오타 수정
- 잘못된 주석 수정

morenow98 committed:
Refactor(*): usecase 패키지의 상위에 application을 추가

morenow98 committed:
Test(*): 북마크 목록 조회의 추가적인 테스트 코드 작성

- 아래의 경우를 테스트하는 코드 작성
1. 멤버가 현재 자격증을 가지고 있지 않을 경우
2. examId와 subjectId가 올바르지 않을 경우

morenow98 committed:
Feat(Problem): 북마크 조회 시 examId 존재에 따라 다른 쿼리가 나가도록 수정

morenow98 committed:
Test(Problem): examId 존재에 따라 다른 쿼리가 나가는 것을 반영한 테스트 코드 구현

- @BeforeEach 사용

morenow98 committed:
Fix(Problem): 테스트 코드가 코드 변경을 반영하지 않았던 것을 수정

morenow98 committed:
Feat(*): 파라미터 검증 로직 수정

- HandlerMethodValidationException 핸들링하는 ExceptionHandler 수정
- MethodArgumentNotValidException 핸들링 메소드 삭제
- COM_001을 invalid request body로, COM_003을 invalid request parameter로 수정

morenow98 committed:
Refactor(Problem): 북마크 조회 시 optional.map 을 사용하도록 변경

morenow98 committed:
Test(Problem): 북마크 조회 테스트 성공 케이스 추가

morenow98 committed:
Feat(Problem): 현재 자격증 로직 변경

- 현재 자격증 검사하는 메소드를 따로 분리
- 현재 자격증 조회 시 if 로직을 dto에서 usecase로 변경

InHyeok-J reviewed (approved):
Member POJO Test 하면 더 좋을거 같음 고생했으

morenow98 commented:
> Member POJO Test 하면 더 좋을거 같음 고생했으

해서 올렸음~!
